go/ai/generate.go

```diff
@@ -86,8 +86,131 @@ func LookupModel(provider, name string) *Model {
 	return (*Model)(core.LookupActionFor[*GenerateRequest, *GenerateResponse, *GenerateResponseChunk](atype.Model, provider, name))
 }
 
-// Generate applies the [Model] to some input, handling tool requests.
-func (m *Model) Generate(ctx context.Context, req *GenerateRequest, cb ModelStreamingCallback) (*GenerateResponse, error) {
+type GenerateParams struct {
+	Request *GenerateRequest
+	Stream  ModelStreamingCallback
+}
+
+// GenerateParamsBuilder adds.
+type GenerateParamsBuilder func(req *GenerateParams)
+
+// WithTextPrompt adds a simple text user prompt to GenerateRequest.
+func WithTextPrompt(prompt string) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Request.Messages = append(req.Request.Messages, NewUserTextMessage(prompt))
+	}
+}
+
+// WithSystemPrompt adds a simple text system prompt to GenerateRequest.
+func WithSystemPrompt(prompt string) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Request.Messages = append(req.Request.Messages, NewSystemTextMessage(prompt))
+	}
+}
+
+// WithMessages adds provided messages to GenerateRequest.
+func WithMessages(messages ...*Message) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Request.Messages = append(req.Request.Messages, messages...)
+	}
+}
+
+// WithConfig adds provided config to GenerateRequest.
+func WithConfig(config any) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Request.Config = config
+	}
+}
+
+// WithCandidates adds provided candidate count to GenerateRequest.
+func WithCandidates(c int) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Request.Candidates = c
+	}
+}
+
+// WithContext adds provided context to GenerateRequest.
+func WithContext(c ...any) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Request.Context = c
+	}
+}
+
+// WithTools adds provided tools to GenerateRequest.
+func WithTools(tools ...Tool) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		var toolDefs []*ToolDefinition
+		for _, t := range tools {
+			toolDefs = append(toolDefs, t.Definition())
+		}
+		req.Request.Tools = toolDefs
+	}
+}
+
+// WithOutputSchema adds provided output schema to GenerateRequest.
+func WithOutputSchema[T any](schema T) GenerateParamsBuilder {
```

Comment:
Does `[T any]` do anything here? Please explain in the comment that the arg is the struct/value, not the schema map or anything else.

Reply:
hmmm... no... vestige of some previous iteration. Removed.